src/api-engine/api/routes/chaincode/views.py

```diff
@@ -92,76 +92,74 @@ def list(self, request):
-                                found = True
-                                os.chdir(cwd)
-                                break
-                # if can not find go.mod, use the dir after extract zipped_file
-                if not found:
-                    for _, dirs, _ in os.walk(file_path):
-                        chaincode_path = file_path + "/" + dirs[0]
-                        break
 
                 org = request.user.organization
                 qs = Node.objects.filter(type="peer", organization=org)
                 if not qs.exists():
-                    raise ResourceNotFound
+                    return Response(
+                        err("At least 1 peer node is required for the chaincode package upload."), 
+                        status=status.HTTP_400_BAD_REQUEST
+                    )
                 peer_node = qs.first()
                 envs = init_env_vars(peer_node, org)
-
                 peer_channel_cli = PeerChainCode("v2.2.0", **envs)
-                res = peer_channel_cli.lifecycle_package(
-                    name, version, chaincode_path, language)
-                os.system("rm -rf {}/*".format(file_path))
-                os.system("mv {}.tar.gz {}".format(name, file_path))
-                if res != 0:
-                    return Response(err("package chaincode failed."), status=status.HTTP_400_BAD_REQUEST)
+                return_code, content = peer_channel_cli.lifecycle_calculatepackageid(temp_path)
+                if (return_code != 0):
+                    return Response(
+                        err("Calculate packageid failed for {}.".format(content)), 
+                        status=status.HTTP_400_BAD_REQUEST
+                    )
+                packageid = content.strip()
+
+                # check if packageid exists
+                cc = ChainCode.objects.filter(package_id=packageid)
+                if cc.exists():
+                    return Response(
+                        err("Packageid {} already exists.".format(packageid)), 
+                        status=status.HTTP_400_BAD_REQUEST
+                    )
+
+                # try to save chaincode package
+                ccpackage_path = os.path.join(FABRIC_CHAINCODE_STORE, packageid)
+                if not os.path.exists(ccpackage_path):
+                    os.makedirs(ccpackage_path)
+                # tared_file = tarfile.TarFile(temp_path)
+                with tarfile.open(temp_path) as tared_file:
+                    tared_file.extractall(ccpackage_path)
+
+                with open(os.path.join(ccpackage_path, "metadata.json"), 'r') as f:
+                    metadata = json.load(f)
+                    language = metadata["type"]
+                    label = metadata["label"]
+
+                os.system("rm -rf {}/*".format(ccpackage_path))
+
+                ccpackage = os.path.join(ccpackage_path, file.name)
+                shutil.copy(temp_path, ccpackage)
```

Comment:
can we execute file read operation in a different process?

Reply:
now start a thread for read and extraction